core/src/main/java/org/axonframework/serialization/compression/AbstractSerializerCompressionWrapper.java

```diff
@@ -0,0 +1,116 @@
+package org.axonframework.serialization.compression;
+
+import org.axonframework.serialization.*;
+
+/**
+ * Wrapper around an embedded {@link Serializer} that can serialize to a byte array. Implementations must implement the
+ * {@code doCompress} and {@code doDecompress} methods. When a NotCompressedException occurs the {@code deserialize}
+ * method tries to deserialize using the {@code embeddedSerializer} without decompression.
+ *
+ * @author Michael Willemse
+ */
+public abstract class AbstractSerializerCompressionWrapper implements Serializer {
+
+    protected Serializer embeddedSerializer;
+
+    /**
+     * Initializes AbstractByteArrayCompressionSerializer implementation with the embedded serializer as parameter
+     * @param embeddedSerializer serializer that must be able to serialize to a byte array otherwise the constructor
+     *                           throws an IllegalArgumentException.
+     */
+    public AbstractSerializerCompressionWrapper(Serializer embeddedSerializer) {
+        if(!embeddedSerializer.canSerializeTo(byte[].class)) {
+            throw new IllegalArgumentException("Expecting serializer that serializes to byte array.");
+        }
+        this.embeddedSerializer = embeddedSerializer;
+    }
+
+    @Override
+    public <T> SerializedObject<T> serialize(Object object, Class<T> expectedRepresentation) {
+        SerializedObject<T> embeddedSerializedObject = embeddedSerializer.serialize(object, expectedRepresentation);
+        if(byte[].class.equals(expectedRepresentation)) {
+            try {
+                return new SimpleSerializedObject<>(
+                        (T) doCompress((byte[]) embeddedSerializedObject.getData()),
+                        embeddedSerializedObject.getContentType(),
+                        embeddedSerializedObject.getType());
+            } catch (Throwable e) {
+                throw new SerializationException("Unhandled Exception while compressing.", e);
+            }
+        } else {
+            return embeddedSerializedObject;
+        }
+    }
+
+    @Override
+    public <T> boolean canSerializeTo(Class<T> expectedRepresentation) {
+        return embeddedSerializer.canSerializeTo(expectedRepresentation);
+    }
+
+    @Override
+    public <S, T> T deserialize(SerializedObject<S> serializedObject) {
+        if(byte[].class.equals(serializedObject.getContentType())) {
```

Comment:
Actually, the content type could be different formats, and still be 'compatible'. InputStream is an example hereof. However, when using the Serializer to send data through gRPC, a content type of `com.google.protobuf.ByteString` would also be possible.

Serializers should use ContentTypeConverters to convert the incoming format to their working format before deserializing and their working format to the expected format when serializing.
Obviously, a serializing wrapper would not be able to convert to Object types, such as Json Node or a Dom4J Document.

You can construct a `ChainingConverter` instance (no parameters required) to convert between expectedRepresentation types.

The `if(byte[].class.equals(...))` would then become `if (converter.canConvert(serializedObject.getContentType()), byte[].class)`.

Reply:
Given the logic in the convert method the ChainingConverter can't be used when you have converters where both source and target type are the same (byte[])

``` java
public <T> T convert(Object original, Class<?> sourceType, Class<T> targetType) {
        if (sourceType.equals(targetType)) {
            return (T) original;
        }
        ...
}
```
Potentially we could use 2 separate ContentTypeConverters and use those in the abstract class (compress/decompress) ?